include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+//
+// Copyright (c) 2017 Damian Jarek (damian dot jarek93 at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/djarek/gabia
+//
+
+#ifndef GABIA_SECURE_SOCKET_HPP
+#define GABIA_SECURE_SOCKET_HPP
+
+#include <gabia/bytes.hpp>
+#include <gabia/crypto/aead.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#include <boost/endian/arithmetic.hpp>
+
+#define CORO_YIELD BOOST_ASIO_CORO_YIELD
+#define CORO_REENTER BOOST_ASIO_CORO_REENTER
+
+namespace gabia {
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+
+template <typename Stream>
+class secure_socket {
+public:
+    using next_layer_type = Stream;
+
+    constexpr static auto max_payload_length = 1024;
+
+    secure_socket(asio::io_service& service) : next_layer_stream{service} {}
+
+    secure_socket(next_layer_type&& next_layer_stream,
+                  crypto::aead_secrets secrets)
+        : next_layer_stream{std::move(next_layer_stream)}, secrets{secrets} {}
```

Comment:
It is necessary to support any user-defined type for the `Stream` template parameter, including references. For example, if you want it to work with this class:
https://github.com/boostorg/beast/blob/20a8f7d75b45f0ababc2547f50833b7f4ba636d3/test/extras/include/boost/beast/test/fail_stream.hpp#L60

Example of use:
```
beast::test::fail_counter fc{n};
boost::asio::io_service ios;
secure_socket<beast::test::fail_stream<tcp::socket>> stream{fc, ios};
```